x/twap/listeners_test.go

```diff
@@ -77,58 +78,222 @@ func (s *TestSuite) TestAfterPoolCreatedHook() {
+			s.Require().NoError(err)
+
+			// run basic swap on the first block if set true
+			if tc.block1Swap {
+				s.RunBasicSwap(poolId)
+			}
+
+			// check that we have correctly stored changed pools
+			changedPools := s.twapkeeper.GetChangedPools(s.Ctx)
+			s.Require().Equal(1, len(changedPools))
+			s.Require().Equal(poolId, changedPools[0])
+
+			s.EndBlock()
+			s.Commit()
+
+			// Second block
+			secondBlockTime := s.Ctx.BlockTime()
+
+			// get updated twap record after end block
+			twapAfterBlock1, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), denom0, denom1)
+			s.Require().NoError(err)
+
+			// if no swap happened in block1, there should be no change
+			// in the most recent twap record after epoch
+			if !tc.block1Swap {
+				s.Require().Equal(twapAfterPoolCreation, twapAfterBlock1)
+			} else {
+				// height should not have changed
+				s.Require().Equal(twapAfterPoolCreation.Height, twapAfterBlock1.Height)
+				// twap time should be same as previous blocktime
+				s.Require().Equal(twapAfterPoolCreation.Time, baseTime)
+
+				// accumulators should not have increased, as they are going through the first epoch
+				s.Require().Equal(sdk.ZeroDec(), twapAfterBlock1.P0ArithmeticTwapAccumulator)
+				s.Require().Equal(sdk.ZeroDec(), twapAfterBlock1.P1ArithmeticTwapAccumulator)
+			}
+
+			// check if spot price has been correctly updated in twap record
+			asset0sp, err := s.App.GAMMKeeper.CalculateSpotPrice(s.Ctx, poolId, twapAfterBlock1.Asset0Denom, twapAfterBlock1.Asset1Denom)
+			s.Require().NoError(err)
+			s.Require().Equal(asset0sp, twapAfterBlock1.P0LastSpotPrice)
+
+			// run basic swap on block two for price change
+			if tc.block2Swap {
+				s.RunBasicSwap(poolId)
+			}
 
-	s.Require().Equal([]uint64{poolId}, s.twapkeeper.GetChangedPools(s.Ctx))
+			s.EndBlock()
+			s.Commit()
+
+			// Third block
+			twapAfterBlock2, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), denom0, denom1)
+			s.Require().NoError(err)
+
+			// if no swap happened in block 3, twap record should be same with block 2
+			if !tc.block2Swap {
+				s.Require().Equal(twapAfterBlock1, twapAfterBlock2)
+			} else {
+				s.Require().Equal(secondBlockTime, twapAfterBlock2.Time)
+
+				// check accumulators incremented - we test details of correct increment in logic
+				s.Require().True(twapAfterBlock2.P0ArithmeticTwapAccumulator.GT(twapAfterBlock1.P0ArithmeticTwapAccumulator))
+				s.Require().True(twapAfterBlock2.P1ArithmeticTwapAccumulator.GT(twapAfterBlock1.P1ArithmeticTwapAccumulator))
+			}
+
+			// check if spot price has been correctly updated in twap record
+			asset0sp, err = s.App.GAMMKeeper.CalculateSpotPrice(s.Ctx, poolId, twapAfterBlock1.Asset0Denom, twapAfterBlock2.Asset1Denom)
+			s.Require().NoError(err)
+			s.Require().Equal(asset0sp, twapAfterBlock2.P0LastSpotPrice)
+		})
+	}
```

Comment:
Do you think it is possible / makes sense to move this into a helper function? It seems that a lot of the logic is shared between swap1 and swap2 so we could potentially avoid code duplication

Reply:
Hrmm I think this might be rather more confusing to abstract it into a helper function since there are alot of parameters involved in this function / testing block, the helper function would have to take in alot of inputs, which I don't think is worth the trade off here imo. Lmk what you think!